crates/node/src/scheduler/mod.rs

```diff
@@ -223,7 +232,117 @@ impl Scheduler {
+                        continue; // All good
+                    }
+                    Ok(false) => {
+                        // VM is not running anymore.
+                        tracing::warn!(
+                            "VM {} running for task (id: {}, tx hash: {}) has stopped",
+                            task.vm_id,
+                            task.task.id,
+                            task.task.tx
+                        );
+                        zombies.push((task.task.id, task.vm_id.clone()));
+                    }
+                    Err(err) => {
+                        tracing::error!("failed to query VM state for {} running task (id: {}, tx hash: {}): {}", task.vm_id, task.task.id, task.task.tx, err);
+                        zombies.push((task.task.id, task.vm_id.clone()));
+                    }
+                }
+            } else {
+                // VM doesn't exist. Shouldn't happen...
+                tracing::error!("found task (id: {}, tx hash: {}) running on VM {} but not the corresponding running VM", task.task.id, task.task.tx, task.vm_id);
+                zombies.push((task.task.id, task.vm_id.clone()));
+            }
+        }
+
+        while let Some((task_id, vm_id)) = zombies.pop() {
+            tracing::debug!(
+                "reaping stopped task (id: {}) running on VM {}",
+                task_id,
+                vm_id
+            );
+
+            // For each zombie task:
+            // - Stop the VM.
+            // - Remove the VM from the `running_vms`.
+            // - Remove the task from the `running_tasks`.
+            //
+            if let Some(idx) = running_vms.iter().position(|x| x.vm_id().eq(vm_id.clone())) {
+                let vm_handle = running_vms.swap_remove(idx);
+                if let Err(err) = self
+                    .program_manager
+                    .lock()
+                    .await
+                    .stop_program(vm_handle)
+                    .await
+                {
+                    tracing::error!("failed to stop VM {}: {}", vm_id, err);
+                }
+            }
+
+            if let Some(idx) = running_tasks.iter().position(|x| x.task.id == task_id) {
+                let _ = running_tasks.swap_remove(idx);
+            }
+        }
+
+        // Now, reap VMs that don't have running task.
+        let mut zombies = vec![];
+        for vm_handle in running_vms.iter() {
+            if !running_tasks.iter().any(|x| x.vm_id.eq(vm_handle.vm_id())) {
+                // XXX: Following is not 100% correct. It checks the runtime
+                // from the very beginning of the whole VM's start-up, which
+                // is not the same as idle runtime. However, right now each VM
+                // is expected to be run only for one task and this check is
+                // not visited until there's no running task for the given VM
+                // so it's ought to be ok.
```

Comment:
This requires some scheduler-internal book keeping for per-VM idle time. I didn't add it for now as we don't really have problem with idle VMs given that those where program fails to start are reaped soon enough.